### DISCUSSION TITLE
Gleichzeitige Verwendung von mehreren Leseköpfen an IR Schnittstelle

### DISCUSSION THREAD

**Discussion by mslookup** [micw/tibber-pulse-reader#8] [General]
Hallo zusammen,

in Issue https://github.com/micw/tibber-pulse-reader/issues/5 hat sich eine kleine Diskussion hinsichtlich des gleichzeitigen Betriebes von mehreren Leseköpfen entwickelt.
Da bereits mehrere Wortbeiträge existieren und das eigentliche Thema in dem ursprünglichen Issue ein anderes ist, habe ich gedacht, wir starten hier eine neue Diskussion und haben damit einen Platz, um uns ausgiebig über das Thema zu unterhalten.

Ich selbst habe auch bereits meine Erfahrungen gemacht und würde die ebenfalls gerne mit euch teilen.

**Comment by micw** [comment]
Dann stelle ich hier gleich nochmal mein Hardware-Projekt vor:

https://u.easyeda.com/account/user/projects/index/detail?project=7ab44df074294c088a309550d36052bf&folder=all

![grafik](https://user-images.githubusercontent.com/186753/230054793-cd38f15e-0d6a-4519-9f08-613ad389a471.png)

Das ist ein einfacher, preiswerter Lese-Schreibkopf mit UART auf 3,3 oder 5 Volt. IR-Transistor und/oder IR-Diode müssen noch eingelötet werden, je nach dem ob er empfangen oder senden kann. Man kann bei 2 der Köpfe RX/TX über kreuz verbinden, dann gibt der eine Kopf auf der TX Seite genau das aus, was bei dem anderen auf der RX-Seite reinkommt. Auch mehrere TX Köpfe können so verbunden werden.

Zusätzlich kann man an RX/TX einen billigen USB-TTL-Adapter anschließen und die Daten direkt per USB mitlesen. Alternativ (oder zusätzlich) kann man auch einen ESP8266 oder anderen µC (z.B. Raspberry) anschließen und die Daten per serieller Schnittstelle mitlesen.

Die Schaltung läuft bei mir mit folgendem Testaufbau:
* Leseeinheit am Zähler incl. der Kontroll-LED, die das sonst unsichtbare IR-Signal visualisiert
* IR-Diode sendet an einen Pulse, der daneben liegt
* Daten werden zusätzlich von einem ESP8266 abgegriffen, decodiert und Zählerstand und Leistungswerte per MQTT ausgegeben

Parallel habe ich die Platinen bei JLPCB bestellt. Die Produktion wurde heute abgeschlossen, im Laufe der kommenden Woche werden sie hier eintreffen und ich kann sie prüfen.

**Reply by mslookup** [reply]
Klingt super, wenn das Teil am Ende so funktioniert, wie es du dir erhoffst, dann kann ich ggf. eine passendes Case entwerfen und mit meinem 3D Drucker drucken...

**Reply by micw** [reply]
Super. Ein Case-Design könnte ich auch direkt mit bei JLL drucken lassen. Kann nur keins entwerfen ^^

Ein Mega Vorteil eines Case wäre, dass der Abstand zu den LEDs größer sein köntne und ich die auf die Rückseite löten könnte (ohne das Loch zum durchschieben)

**Reply by mslookup** [reply]
Ich bin auf jedenfall sehr gespannt, was dabei am Ende rauskommt.

Ideen für ggf. spätere Revisionen: 
- Als Datenabgreifewerkzeug bietet sich ein ein ESP01s an, der hat den genannten ESP8266 und funktioniert bei mir hervorragend. Wird auch in dem von Hichi verkauften All-in-One Reader eingesetzt und kann mit einem 8-poligen Stecker direkt mit einer Platine verbunden werden, wenn man dort eine entsprechende Schnittstelle vorsieht.
- Als mögliche Stromquelle, die man einfach ein und ausstecken kann, wäre ggf. eine USB-C / Mikro-USB Schnittstelle denkbar.

**Reply by micw** [reply]
Eine schöne Idee. Für den ESP benötigt man bei USB aber noch einen Spannungsregler (bei der Leistung sollte ein Linearregler wie AMS1117 reichen). Damit wird's schon schnell eng auf dem Board, insbesondere wenn die Lösung flexibel einsetzbar bleiben soll (z.B. will ich das gleiche Board als Sender für den Pulse nehmen, welches ich als Empfänger für den Zähler habe - dort will ich aber weder Spanungsregler noch ESP drauf haben).

Meine Idee ist, da eher ein größeres Support-Board zu machen mit Optionen für verschiedene Bestückungen.

Hab da schon mal was angefangen:

![grafik](https://user-images.githubusercontent.com/186753/230154425-caae2f76-e115-4954-af81-abbd7c204d5a.png)

* bis zu 4 Köpfe
* 3.3V Eingang oder "sonstiger" Eingang mit Buck/Boost Spannungsregler
* D1 Mini ESP8266
* USB-TTL-Adapter
* alles optional, je nach Bedarf

**Reply by mslookup** [reply]
Sieht gut aus, auch wenn ich zum aktuellen Zeitpunkt nur gut 80% davon technisch verstehe 👍 
Meine Punkte waren natürlich nur Vorschläge, die ich mal irgendwo niederschreiben wollte. 
Ich persönlich wäre/bin mit allem zufrieden, was funktioniert :)

**Comment by mslookup** [comment]
Ich würde auch gerne meine bisherigen Versuche vorstellen:

Ich hatte mir bei Hichi 3 fertig gelötete TTL IR Leseköpfe (https://www.ebay.de/itm/314015558154) bestellt, um mit einem die Daten vom Zähler zu empfangen und die selben Daten mit den anderen beiden wieder auszusenden. 5V Spannungsquelle war ein USB Netzteil.

Alle 3 Leseköpfe habe ich mit jeweils 4 Jumperkabel belötet. Im Anschluss habe ich alle 3 an die Spannungsquelle mit + und - an das USB Kabel angeschlossen, TX des Lesekopfes mit den 2 RX der Sendeköpfe und die RX des Lesekopfes auf die 2 TX der Sendeköpfe geschlossen. Den Lesekopf an den Stromzähler drangehängt und an den Sendeköpfen das schwache Pulsieren der Sende-LEDs beobachtet. Soweit so gut. Wenn ich jetzt allerdings meinen fertigen Hichi mit ESP01S und Tasmota an die Sendeköpfe dranhalte, empfängt er leider nichts verwertbares. Mit dem Tibber Pulse ist es ähnlich - er blinkt zwar putz munter weiß an der LED, leider zeigt er mir in der App oder über MQTT keine Daten an. Habe dann mehrere Stunden rumprobiert und auch 3D Gehäuse gedruckt, damit die Lese und Sendedioden auf den IR-Köpfen sich nicht mehr sehen - Lötstellen mit Messgeräten geprüft, usw. Leider konnte ich das Problem nicht ausmachen und habe schließlich den Versuch eingestampft.

Ich bin absolut kein Mikroelektronik Guru, aber nach meinem Verständnis nach ist das genau das, was micw auch gemacht hat, nur in einer abgewandelten Form. Ich habe dazu auch mal bei Hichi direkt nachgefragt, er hat mir ein Youtube Video (https://youtu.be/YY7A3pxg81s) gesendet, in dem er zeigt, dass das ganze grundsätzlich funktionieren sollte. Auf meine Frage, ob er nicht an einer verkaufbaren Produktneuentwicklung interessiert sei, ist er nicht eingegangen :)

Nach meinem Fehlschlag hab ich mir gedacht: "Kann ja nicht sein...", also habe ich direkt an meinen funktionierenden Hichi, an den ESP01S, an TX (getauscht), RX(getauscht), GND und 3,3V einen der 3 TTL IR Köpfe angelötet. Ergebnis war das selbe wie beim vorherigen Versuch: Die Sendediode hat fleißig leicht rötlich immer wieder gezuckt, aber der Pulse konnte keine brauchbaren Daten in dem Zucken der Sendediode erkennen. Seitdem habe ich das experimentieren eingestellt und hoffe auf eine Ready2Use-Lösung von z.B. euch. 

Nachdem meine Mikroelektronik und Lötkenntnisse anscheinend nicht so gut zu gebrauchen sind, kann ich anbieten, wenn jemand was funktionierendes, Plug and Playfähiges hat, das gerne von Ihm käuflich zu erwerben und ggf. ein Case dafür zu designen und mit dem 3D Drucker auszudrucken.

VG mslookup

**Reply by micw** [reply]
Interssanter Weise hatte ich für TX erst auch eine Schaltung mit einem PNP Transistor (das ist die selbe Schaltung, die ich jetzt für die Kontroll-LED verwende). Obwohl es von der Elektronik sauber gehen müsste, hatte ich den selben, von Dir beschriebenen Effekt. Ich vermute, dass das Signal zu schwach oder die Signalflanken nicht sauber genug sind. Ich habe am Montag mein Oszilloskop bekommen und schau mir das nochmal an.

Inzwischen hatte ich die Schaltung mit den 2 Transistoren. Der erste invertiert das Signal (notwendig für UART), der zweite verstärkt es für die LED. Damit funktioniert es problemlos. LED muss nur grob in Richtung Pulse zeigen und schon kommen die Daten an.

Die Schaltung ist nicht so elegant wie die mit einem Transistor aber da ein weiterer Transistor nur ein paar Cent kostet, habe ich die für meine Platine übernommen.

**Reply by mslookup** [reply]
Krass, dass du auch an dem Punkt warst :) Ich dachte schon, ich hätt mich da irgendwo verlötet^^
Auf laaaange Sicht plane ich mir das Wissen so anzueignen, dass ich einen Adapter ggf. selbst bauen kann. 
Meine bevorzugte Lösung ist eher kompakt und geht in Richtung "Erweiterung des 'Hichi IR WLAN' um eine Platine, die man einfach auf den ESP01s des Hichi drauflöten/draufsetzen kann".  An dem verbauten ESP01s des Hichi liegen bereits 3,3V, GND (beides über USB) sowie TX und RX an. Ob das ganze technisch/physikalisch überhaupt funktionieren kann, mag ich zur Zeit nicht beurteilen. Dazu fehlt mir das Wissen über die Funktionsweise solcher Mikroelektronik im Generellen.

**Reply by micw** [reply]
Im Prinzip müsstest Du meinen Schreib-Lesekopf mit einer IR-LED auf der Schreibseite bestückt an 3V3, GND und mit TX an deinem RX anschließen, dann müsste es gehen - wenn bei dir wirklich TTL UART rauskommt

Weißt du, ob die Daten von Deinem Zähler an RX oder TX anliegen (das ist ja immer abhängig von der Perspektive, was RX und TX ist)? Wenn ja, miss mal von da nach GND. Dort müssten fast immer 3V3 anliegen, außer der Zähler sendet. Wenn es umgekehrt ist, dann ist es kein UART sondern inverses UART.

**Reply by mslookup** [reply]
Ich hab mal nachgemessen: TX zeigt durchgängig 3,3V ohne Dynamik. Bei RX sieht es anders aus - 3,3V, außer alle 1-2 Sek., da geht die Spannung für einen kurzen Zeitraum runter. 
Somit gehe ich gehe davon aus, dass die Daten des Zählers am ESP01s an RX anliegen. Nach deiner Beschreibung müsste es UART sein.

**Reply by micw** [reply]
Schaut nach UART auf 3.3V-Level aus. Das Signal könntest du in einen UART-TTL USB Adapter, direkt in einen ESP8266 oder ESP32 oder in den 3.3V seriellen Eingang eines Raspberry Pi einspeisen und würdest sehr wahrscheinlich die SML Daten bei 9600 baud sehen. Meine Platine kannst Du sehr wahrscheinlich daran anschließen und den Pulse dran tun und es geht.

**Reply by micw** [reply]
Wenn du 2 passende Transistoren, die 3 Widerstände und eine IR-LED hast, kannst du es auch in fliegernder Verkabelung ausprobieren.

**Comment by micw** [comment]
Die Diskussion wurde in https://github.com/evcc-io/evcc/discussions/7070 fortgesetzt.
